
# Bioinformatics softwares
**Tips:**
- Read the official article about how the program works, and then read the manual on the official website for the command for running the program.

## Quality control tools
### FastQC 
For official website, go to [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). <br>
For manual, type ``fastqc -h``. 

### MultiQC
For official website, go to [MultiQC](https://multiqc.info/). <br>
For official article, read [Philip Ewels and others, MultiQC: summarize analysis results for multiple tools and samples in a single report, Bioinformatics, Volume 32, Issue 19, October 2016, Pages 3047–3048](https://academic.oup.com/bioinformatics/article/32/19/3047/2196507).

### Qualimap
For official website, go to [Qualimap 2.3 documentation](http://qualimap.conesalab.org/doc_html/index.html). <br>
For official article, go to [Fernando García-Alcalde and others, Qualimap: evaluating next-generation sequencing alignment data, Bioinformatics, Volume 28, Issue 20, October 2012, Pages 2678–2679](https://academic.oup.com/bioinformatics/article/28/20/2678/206551).

## Trimming tools
### Trimmomatic
For official website, go to [Trimmomatic: A flexible read trimming tool for Illumina NGS data](http://www.usadellab.org/cms/?page=trimmomatic). <br> 
For official article, read [Anthony M. Bolger and others, Trimmomatic: a flexible trimmer for Illumina sequence data, Bioinformatics, Volume 30, Issue 15, August 2014, Pages 2114–2120](https://academic.oup.com/bioinformatics/article/30/15/2114/2390096). 

## Genome assembler
### SPAdes
For official website, go to [SPAdes](https://cab.spbu.ru/software/spades/). <br>
Alekseyev M. A. SPAdes: новый ассемблер геномов с поддержкой одноклеточного секвенирования. (talk), 2011. <br>
For official artical, read [Bankevich A, Nurk S, Antipov D, Gurevich AA, Dvorkin M, Kulikov AS, Lesin VM, Nikolenko SI, Pham S, Prjibelski AD, Pyshkin AV, Sirotkin AV, Vyahhi N, Tesler G, Alekseyev MA, Pevzner PA. SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. J Comput Biol. 2012 May;19(5):455-77](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3342519/). 

## Quality assessment for genome assembly
### Quast
For official website, go to [Quast](https://quast.sourceforge.net/). <br>
For official article, read [Alexey Gurevich and others, QUAST: quality assessment tool for genome assemblies, Bioinformatics, Volume 29, Issue 8, April 2013, Pages 1072–1075](https://academic.oup.com/bioinformatics/article/29/8/1072/228832).

### Bandage
For official website, go to [Bandage](http://rrwick.github.io/Bandage/). <br>
For official article, read [Ryan R. Wick and others, Bandage: interactive visualization of de novo genome assemblies, Bioinformatics, Volume 31, Issue 20, October 2015, Pages 3350–3352](https://academic.oup.com/bioinformatics/article/31/20/3350/196114). 

## Aligners
### Mauve
For official website, go to [mauve](https://darlinglab.org/mauve/mauve.html). <br>
For official article, read [Darling AC, Mau B, Blattner FR, Perna NT. Mauve: multiple alignment of conserved genomic sequence with rearrangements. Genome Res. 2004 Jul;14(7):1394-403](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC442156/). 

### BWA 
For official website, go to [Burrows-Wheeler Aligner](https://bio-bwa.sourceforge.net/). <br>
For official article, read [Heng Li , Richard Durbin, Fast and accurate short read alignment with Burrows–Wheeler transform, Bioinformatics, Volume 25, Issue 14, July 2009, Pages 1754–1760](https://academic.oup.com/bioinformatics/article/25/14/1754/225615?login=false).

## Annotation tools
### Prokka
For official website, go to [Prokka: rapid prokaryotic genome annotation](https://github.com/tseemann/prokka). <br>
For official artical, read [Torsten Seemann, Prokka: rapid prokaryotic genome annotation, Bioinformatics, Volume 30, Issue 14, July 2014, Pages 2068–2069](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517).

## Gap fillers
### GapFiller
An earlier artical, [Nadalin F, Vezzi F, Policriti A: GapFiller: a preprocessing step for the de novo assembly problem [abstract]. Proceedings on the 8th annual meeting of the Bioinformatics Italian Society. 2011, 13-14](http://bioinformatics.hsanmartino.it/bits_library/library/01025.pdf). <br>
For official artical, read [Nadalin, F., Vezzi, F. & Policriti, A. GapFiller: a de novo assembly approach to fill the gap within paired reads. BMC Bioinformatics 13 (Suppl 14), S8 (2012)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S14-S8). 

### IMAGE
For official article, read [Tsai IJ, Otto TD, Berriman M. Improving draft assemblies by iterative mapping and assembly of short reads to eliminate gaps. Genome Biol. 2010;11(4):R41](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2884544/). 

# Core genome identifier 
### Spine 
For offical website, go to [Spine](https://github.com/egonozer/Spine). <br>
For official article, go to [Ozer, E.A., Allen, J.P. & Hauser, A.R. Characterization of the core and accessory genomes of Pseudomonas aeruginosa using bioinformatic tools Spine and AGEnt. BMC Genomics 15, 737 (2014)](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-737#citeas). 
